src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,234 @@
+using System;
+
+namespace D2L.CodeStyle.Analyzers.Async {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class BlockingAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Diagnostics.AsyncMethodCannotBeBlocking,
+			Diagnostics.AsyncMethodCannotCallBlockingMethod,
+			Diagnostics.OnlyCallBlockingMethodsFromMethods,
+			Diagnostics.BlockingCallersMustBeBlocking,
+			Diagnostics.UnnecessaryBlocking
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.ConfigureGeneratedCodeAnalysis(
+				GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics
+			);
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		private static void CompilationStart( CompilationStartAnalysisContext ctx ) {
+			var attr = ctx.Compilation.GetTypeByMetadataName(
+				"D2L.CodeStyle.Annotations.BlockingAttribute"
+			);
+
+			if( attr == null ) {
+				return;
+			}
+
+			var haveBlockingAttribute = new ConcurrentDictionary<IMethodSymbol, bool>(
+				SymbolEqualityComparer.Default
+			);
+
+			ctx.RegisterSymbolAction(
+				ctx => InspectMethodDeclaration(
+					ctx,
+					attr,
+					noteMethodWithBlockingAttribute: m => haveBlockingAttribute[m] = true
+				),
+				SymbolKind.Method
+			);
+
+			var callsSomethingBlocking = new ConcurrentDictionary<IMethodSymbol, bool>(
+				SymbolEqualityComparer.Default
+			);
+
+			ctx.RegisterSyntaxNodeAction(
+				ctx => InspectMethodCall(
+					ctx,
+					attr,
+					noteMethodHasBlockingCall: m => callsSomethingBlocking[m] = true
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			// After the rest of analysis report about potentially unnecessary [Blocking]s
+			ctx.RegisterCompilationEndAction(
+				ctx => WarnForUnusedBlockingAttributes(
+					ctx,
+					attr,
+					haveBlockingAttribute: haveBlockingAttribute.Keys,
+					callsSomethingBlocking: callsSomethingBlocking.Keys
+				)
+			);
+		}
+
+		private static void InspectMethodDeclaration(
+			SymbolAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodWithBlockingAttribute
+		) {
+			// Goal: find async methods that have [Blocking]
+
+			var methodSymbol = (IMethodSymbol)ctx.Symbol;
+
+			if( !methodSymbol.IsAsync ) {
+				// Make note if it has [Blocking], because it may not be necessary.
+				if( HasAttribute( methodSymbol, blockingAttr ) ) {
+					noteMethodWithBlockingAttribute( methodSymbol );
+				}
+
+				return;
+			}
+
+			var attrData = GetAttribute( methodSymbol, blockingAttr );
+
+			if( attrData == null ) {
+				return;
+			}
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.AsyncMethodCannotBeBlocking,
+					attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+					methodSymbol.Name
+				)
+			);
+
+			// TODO: code fix to remove [Blocking]

```

Comment:
I left all the code fixes for the next PR.

Having the fixes will be handy e.g. for switching things to `async` manually/lazily. Mostly there are only a couple of things that can get reused a few times.